### PR TITLE
AEA-1311, AEA-1336 Simplify and streamline message routing

### LINKED ISSUE #1922: Routing map from connections to skills
**Is your feature request related to a problem? Please describe.**
Currently, default routing only applies for routing from skills to connections. When having multiple connections processing the same protocol (e.g. http_client and http_server) it can be helpful to specify which connection maps to which skill.

**Describe the solution you'd like**
Solution to above problem

### PR DESCRIPTION
**Pull request by DavidMinarsch**
## Proposed changes

This PR improves message routing by doing the following:
- introduce handling of public id based routing in `to` field of envelopes/messages
- removing EnvelopeContext where it is not needed
- clarifying routing rules in docs

## Fixes

#1922, #2036

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

-

### CODE REVIEW

File: docs/message-routing.md [hunk 1]
```diff
@@ -3,23 +3,28 @@ Message routing can be split up into the routing of incoming and outgoing `Messa
...
 ## Incoming `Messages`
 
-- `Connections` receive `Envelopes` which they deposit in `InBox`
-- `AgentLoop`'s react picks `Envelopes` off the `InBox`
-- the AEA tries to decode the message; errors are handled by the error `Skill`
-- `Messages` are dispatched to all relevant `Handlers`. To limit dispatch to a specific `Handler` in a specific `Skill` the `EnvelopeContext` can be used to reference a unique `skill id`.
+- `Connections` receive or create `Envelopes` which they deposit in the `InBox`
+- the `AgentLoop` picks `Envelopes` off the `InBox`
+- the `AEA` tries to decode the message; errors are handled by the `ErrorHandler`
+- `Messages` are dispatched based on two rules:
```
Comment (DavidMinarsch) [inline docs/message-routing.md:9]:
```suggestion
- `Messages` are dispatched based on three rules:
```

File: aea/test_tools/test_skill.py [hunk 2]
```diff
@@ -150,6 +151,7 @@ def build_incoming_message(
         target: Optional[int] = None,
         to: Optional[Address] = None,
         sender: Address = COUNTERPARTY_ADDRESS,
+        is_agent_to_agent_messages: Optional[bool] = None,
```
Comment (DavidMinarsch) [inline aea/test_tools/test_skill.py:154]: update docstring

File: docs/message-routing.md [hunk 3]
```diff
@@ -3,23 +3,28 @@ Message routing can be split up into the routing of incoming and outgoing `Messa
 
 ## Incoming `Messages`
```
Comment (DavidMinarsch) [inline docs/message-routing.md:4]: Add explanation on:

- interaction types: agent to agent and component to component and its implications on the message to/sender fields

File: packages/fetchai/connections/http_server/connection.py [hunk 4]
```diff
@@ -166,22 +166,19 @@ async def create(cls, http_request: BaseRequest) -> "Request":
...
-        :param connection_id: id of the connection
-        :param agent_address: agent's address
         :param dialogue_reference: new dialog refernece for envelope
+        :param target_skill_id: the target skill id
 
         :return: envelope
         """
         url = self.full_url_pattern
-        uri = URI(self.full_url_pattern)
-        context = EnvelopeContext(connection_id=connection_id, uri=uri)
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http_server/connection.py:182]: This was somewhat problematic. Firstly, an external entity should not be able to target a skill. Second, if we don't know the target skill we cannot maintain a proper bilateral interaction as we don't know the address (using agent address is not ok as the HTTPDialogues class in the skill will be instantiated with the skill_id as an address - cf. component to component messaging)

File: packages/fetchai/connections/http_server/connection.py [hunk 5]
```diff
@@ -375,6 +371,7 @@ def __init__(
         super().__init__(address=address, connection_id=connection_id)
         self.host = host
         self.port = port
+        self.target_skill_id = target_skill_id
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http_server/connection.py:374]: Introduction of the target_skill_id means the messages generated by the server are now exclusively dealt with by a single skill.

File: packages/fetchai/connections/webhook/connection.yaml [hunk 6]
```diff
@@ -8,13 +8,14 @@ aea_version: '>=0.10.0, <0.11.0'
...
   __init__.py: QmWUKSmXaBgGMvKgdmzKmMjCx43BnrfW6og2n3afNoAALq
-  connection.py: QmYLCDjJLzhXiqTjjSxFUjvTQqhFUU5Qauv8s6M7k5tNT5
+  connection.py: QmSheTDaVWGKq8U8tUGhviS3fz5rYJwZ75tsExEjZYAxaJ
 fingerprint_ignore_patterns: []
 connections: []
 protocols:
 - fetchai/http:0.12.0
 class_name: WebhookConnection
 config:
+  target_skill_id: null
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/webhook/connection.yaml:18]: see http server, same arguments apply here

**Review by jrriehl — APPROVED** [review]
Looks great. Much simpler and the documentation is very helpful.

File: aea/configurations/data_types.py [hunk 7]
```diff
@@ -321,6 +321,21 @@ def from_str(cls, public_id_string: str) -> "PublicId":
...
+
+        :param public_id_string: the public id in string format.
+        :return: the public id object or None
+        """
+        result: Optional[PublicId] = None
+        try:
+            result = cls.from_str(public_id_string)
+        except ValueError:
+            pass
+        return result
```
Comment (marcofavorito) [inline aea/configurations/data_types.py:337]: (minor) A slightly less verbose alternative:
```suggestion
import contextlib

        result: Optional[PublicId] = None
        with contextlib.suppress(ValueError):
            result = cls.from_str(public_id_string)
        return result
```

File: tests/test_packages/test_skills_integration/test_http_echo.py [hunk 8]
```diff
@@ -40,6 +40,10 @@ def test_echo(self):
         self.add_item("connection", "fetchai/http_server:0.16.0")
         self.add_item("skill", "fetchai/http_echo:0.14.0")
         self.set_config("agent.default_connection", "fetchai/http_server:0.16.0")
+        self.set_config(
```
Comment (solarw) [inline tests/test_packages/test_skills_integration/test_http_echo.py:43]: JFYI
found the case when multiple skills are ok to get http server message.
skills can handle various routes/server names, so can filter messages by some criteria.
anyway it's a specific casse

**Pull request merged**